File: typescript/src/process3/activities.ts
import type { ProductChangeSpec, ReviewDecision } from './models';

const defaultSleep = (ms: number) => new Promise((resolve) => setTimeout(resolve, ms));
let sleep = defaultSleep;

export function setActivitySleepForTest(fn: typeof sleep): void {
  sleep = fn;
}

export function resetActivityStateForTest(): void {
  sleep = defaultSleep;
}

export async function notifyReviewer(spec: ProductChangeSpec): Promise<void> {
  await sleep(5000);
  console.log(
    `[NOTIFY] requestId=${spec.requestId} newPrice=${spec.newPrice.toFixed(2)} product=${spec.productId}`
  );
}

export async function recordDecision(
  spec: ProductChangeSpec,
  decision: ReviewDecision
): Promise<void> {
  await sleep(5000);
  console.log(
    `[RECORD] requestId=${spec.requestId}, product=${spec.productId}, review=${decision.outcome}`
  );
}

export async function applyNewPrice(spec: ProductChangeSpec): Promise<void> {
  await sleep(5000);
  console.log(
    `[APPLY] requestId=${spec.requestId} set price=${spec.newPrice.toFixed(2)} for product=${spec.productId}`
  );
}

File: python/src/temporal_learn/process3/activities.py
import asyncio

from temporalio import activity

from .models import ProductChangeSpec, ReviewDecision


# -----------------------------------------------------------------------------
# Activity: notify_reviewer
@activity.defn
async def notify_reviewer(spec: ProductChangeSpec) -> None:
    await asyncio.sleep(5)
    print(
        f"[NOTIFY] requestId={spec.request_id} "
        f"newPrice={spec.new_price:.2f} product={spec.product_id}"
    )


# -----------------------------------------------------------------------------
# Activity: record_decision
@activity.defn
async def record_decision(spec: ProductChangeSpec, decision: ReviewDecision) -> None:
    await asyncio.sleep(5)
    print(
        f"[RECORD] requestId={spec.request_id}, "
        f"product={spec.product_id}, review={decision.outcome.value}"
    )


# -----------------------------------------------------------------------------
# Activity: apply_new_price
@activity.defn
async def apply_new_price(spec: ProductChangeSpec) -> None:
    await asyncio.sleep(5)
    print(
        f"[APPLY] requestId={spec.request_id} "
        f"set price={spec.new_price:.2f} for product={spec.product_id}"
    )
